### ISSUE TITLE
Kotlin objects deallocation

### ISSUE THREAD

**Issue by romanandreyvich** [JetBrains/kotlin-native#4052]
I'm currently trying to move my Swift view models to KotlinNative and here is kind of problem I faced. In short Kotlin objects are not deallocating as I expect.

For example, I have: MyViewController & MyPresenter

MyViewController is shown as modal VC (by the method "present(vc:animated:)") in iOS. 
MyViewController does some job via MyPresenter and shows some UI after presenter's work is finished.
The problem is when I dismiss MyViewController it won't be deallocated because of MyPresenter.
I guess it's happening because I have two strong references and they are refers each other. In pure Swift I'm able to use weak reference to solve the problem, what about KotinNative? Should I Use WeakReference inside KN class? Or it is GarbageCollector issue (because it works slowly then ARC)? Or maybe I do something wrong? :)

Here is an example project: https://drive.google.com/file/d/1NzwMU7QStgQuKBNgty1HZei1ZZvQG5Ys/view?usp=sharing

Thanks in advance!

**Comment by yurii-diachenko** [comment]
Any update in this issue?

**Comment by SvyatoslavScherbina** [comment]
@romanandreyvich 

> I guess it's happening because I have two strong references and they are refers each other.

This definitely might be the reason. Kotlin/Native GC can't handle reference cycles involving Obj-C references, because reference cycles aren't handled by Obj-C ARC in the first place.

> In pure Swift I'm able to use weak reference to solve the problem, what about KotinNative? Should I Use WeakReference inside KN class?

One of these references should be weak. In Kotlin you can use `WeakReference` for this.

> Or it is GarbageCollector issue (because it works slowly then ARC)?

K/N GC doesn't deallocate an object immediately after it becomes unused in any case.

@yurii-diachenko what updates do you expect? Do you encounter the same issue?

**Comment by yurii-diachenko** [comment]
@SvyatoslavScherbina Thank you for an update. You already gave the answer for the issue I was facing with.

> K/N GC doesn't deallocate an object immediately after it becomes unused in any case.

I just noticed that memory dump was keeping all the object instances (when there is no any reference left to these objects) of the class the was declared in a shared Multiplatform module. I have made an experiment and made that shared class to contain a huge String (allocated memory for the object would be about 100Mb). Since object instances grew up in their size - K/N GC started to work and it cleared all unneeded object instances. 

So indeed, K/N works and memory will be cleared. It just won't happen immediately.

Thanks @SvyatoslavScherbina for the answer again
